<a href="https://colab.research.google.com/github/flaviovdf/fcd/blob/main/labs/lab01/sol.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Laboratório 1: Arrays e DataFrames

Bem-vindo ao Laboratório 1! Esta semana, aprenderemos sobre arrays, que nos permitem armazenar sequências de dados, e DataFrames, que nos permitem trabalhar com vários arrays de dados sobre as mesmas coisas. Esses tópicos são abordados em na [aula do DSC10](https://notes.dsc10.com/02-data_sets/arrays.html) nas notas `babypandas`.


**Não use loops `for` em nenhuma pergunta deste laboratório.** Se você não sabe o que é um loop `for`, não se preocupe: ainda não abordamos isso. Mas se você sabe o que eles são e está se perguntando por que não é correto usá-los, é porque os loops em Python são lentos e os loops em arrays e DataFrames geralmente devem ser evitados, já que temos funções nativas de bibliotecas como `numpy` que são muito mais rápidas.

Primeiro, configure as importações necessárias executando a célula abaixo.

In [1]:
# Descomente para executar no Colab
# ! pip install babypandas

In [2]:
import numpy as np
import pandas as pd

Além disso, iremos baixar alguns dados que usaremos ao longo desse notebook. Não se preocupe em entender a célula a seguir, mas basicamente o que estamos fazendo é:

1. Usaremos `wget` que é um comando para baixarmos coisas da internet.
2. Usaremos `unzip` para "dezipar" o arquivo `data.zip`, contendo uma pasta com os nossos dados.
3. Removeremos `data.zip` com o comando `rm` para não ficarmos com arquivos desnecessários.

> Para executarmos comandos de terminal no Jupyter Notebook, iniciamos as linhas com `!`.

In [3]:
#!wget https://github.com/ThiagoPoppe/monitoria_fcd2024/raw/main/labs/lab01/data.zip -P ./
#!unzip data.zip
#!rm data.zip

# 1. Matrizes

Os computadores são mais úteis quando você pode usar uma pequena quantidade de código para *fazer a mesma ação* para *muitas coisas diferentes*.

Por exemplo, no tempo que você leva para calcular a gorjeta de 18% na conta de um restaurante, um laptop pode calcular gorjetas de 18% para cada conta de restaurante paga por cada ser humano na Terra naquele dia. (Isto é, se você for muito rápido em fazer contas de cabeça!)

**Matrizes** são como colocamos muitos valores em um só lugar para que possamos operar neles como um grupo. Por exemplo, se `bilhoes_de_numeros` for uma matriz de números, a expressão

```python
0.18 * bilhoes_de_numeros
```

fornece uma nova matriz de números que é o resultado da multiplicação de cada número em `bilhoes_de_numeros` por 0,18 (18%). As matrizes não estão limitadas a números; também podemos colocar todas as palavras de um livro em uma série de strings.

Concretamente, uma matriz é uma **coleção de valores do mesmo tipo**, como uma coluna em uma planilha (pense no Planilhas Google ou no Microsoft Excel).

<img src="https://github.com/ThiagoPoppe/monitoria_fcd2024/blob/main/labs/lab01/imagens/sheet_array.png?raw=true" width=600>

## 1.1. Fazendo matrizes
Você mesmo pode digitar os dados que vão em um array, mas normalmente não é assim que criaremos arrays. Normalmente, criamos arrays carregando-os de uma fonte externa, como um arquivo de dados.

Porém, primeiro vamos aprender como fazer isso da maneira mais difícil. Para começar, podemos fazer uma **lista** de números colocando-os entre colchetes e separando-os por vírgulas:

In [4]:
minha_lista = [14, -2.26, 0.15]
minha_lista

[14, -2.26, 0.15]

Assim como `int`, `float` e `str`, a `list` é um tipo de dados fornecido pelo Python. As listas são muito flexíveis e fáceis de trabalhar, mas são *lentas* 🐢.

Como cientistas de dados, frequentemente trabalharemos com milhões ou até bilhões de números. Para isso, precisamos de algo mais rápido que uma `lista`. Em vez de listas, usaremos *arrays*.

Arrays são fornecidos por um pacote chamado [NumPy](http://www.numpy.org/) (pronuncia-se "NUM-pai" ou, se você preferir pronunciar as coisas incorretamente, "NUM-pi"). O pacote é chamado `numpy`, mas é padrão abreviá-lo para `np`. Você pode fazer isso com:

```python
import numpy as np
```

Cientistas de dados, bem como engenheiros e cientistas de todos os tipos, usam `numpy` com frequência, e você verá bastante disso se for especialista em ciência de dados.

In [5]:
import numpy as np

Agora, para criar um array, chame a função `np.array` com uma lista de números. Execute esta célula para ver um exemplo:

In [6]:
np.array([14, -2.26, 0.15])

array([14.  , -2.26,  0.15])

Observe que você precisa dos colchetes aqui. Se você tentasse executar o código a seguir, o Python acharia ruim porque você o esqueceu:

```python
np.array(14, -2.26, 0.15)
```

<img src='https://github.com/ThiagoPoppe/monitoria_fcd2024/blob/main/labs/lab01/imagens/brackets.png?raw=true' width=400>

Os próprios arrays também são valores, assim como números e strings. Isso significa que você pode atribuir nomes a eles ou usá-los como argumentos para funções.

**Questão 1.1.1.** Faça um array contendo os números 2, 4 e 6, nessa ordem. Nomeie-o como `numeros_pares`.

In [7]:
numeros_pares = np.array([2,4,6])
numeros_pares

array([2, 4, 6])

**Questão 1.1.2.** Faça um array contendo os números 0, -1, 1, $\pi$ e $e$, nessa ordem. Nomeie-o como `outros_numeros`.

*Dica:* $\pi$ e $e$ estão disponíveis no módulo `np`, que já foi importado. Assim como você usou `math.pi` para obter $\pi$ no último laboratório, você também pode usar `np.pi` para obter $\pi$. **Não** importe o módulo `math`.

In [8]:
outros_numeros = np.array([0, -1, 1, np.pi, np.e])
outros_numeros

array([ 0.        , -1.        ,  1.        ,  3.14159265,  2.71828183])

**Questão 1.1.3.** Faça um array contendo as cinco strings `"Hello"`, `","`, `" "`, `"world"` e `"!"`. (O terceiro é um espaço único entre aspas.) Nomeie-o como `componentes_hello_world`.

*Nota:* Se você imprimir `componentes_hello_world`, você notará algumas informações extras além de seu conteúdo: `dtype='<U5'`. Essa é apenas a maneira extremamente estranha do NumPy de dizer que as coisas no array são strings. Caso você esteja interessado, o `U` significa que esta string está codificada em [unicode](https://en.wikipedia.org/wiki/Unicode), e o `<5` significa que todas as strings no array têm 5 caracteres ou menos.

In [9]:
componentes_hello_world = np.array(["Hello",","," ","World","!"])
componentes_hello_world

array(['Hello', ',', ' ', 'World', '!'], dtype='<U5')

Muitas vezes, em ciência de dados, queremos trabalhar com muitos números espaçados uniformemente dentro de algum intervalo. NumPy fornece uma função especial para isso chamada `arange`. A expressão `np.arange(comeco, fim, espaco)` produz um array com todos os números começando em `comeco`, contando de  `espaco` em `espaco`, parando **antes** de `fim` ser alcançado.

Por exemplo, o valor de `np.arange(1, 8, 2)` é uma matriz com os elementos `1, 3, 5 e 7` - começa em 1 e vai contando de 2 em 2, terminando até chegar no último valor menor que 8. Em outros palavras, ele cria o mesmo array que `np.array([1, 3, 5, 7])`.

`np.arange(4, 9, 1)` é um array com os elementos `4, 5, 6, 7 e 8`, não contendo o 9 porque `np.arange` para *antes* do valor de parada ser atingido.

**Pergunta 1.1.4.** Use `np.arange` para criar um array com todos os múltiplos de 99 de 0 até (**e incluindo**) 9999. (Portanto, seus elementos são 0, 99, 198, 297, etc.)

In [10]:

multiplos_de_99 = np.arange(0,9999,99)
multiplos_de_99

array([   0,   99,  198,  297,  396,  495,  594,  693,  792,  891,  990,
       1089, 1188, 1287, 1386, 1485, 1584, 1683, 1782, 1881, 1980, 2079,
       2178, 2277, 2376, 2475, 2574, 2673, 2772, 2871, 2970, 3069, 3168,
       3267, 3366, 3465, 3564, 3663, 3762, 3861, 3960, 4059, 4158, 4257,
       4356, 4455, 4554, 4653, 4752, 4851, 4950, 5049, 5148, 5247, 5346,
       5445, 5544, 5643, 5742, 5841, 5940, 6039, 6138, 6237, 6336, 6435,
       6534, 6633, 6732, 6831, 6930, 7029, 7128, 7227, 7326, 7425, 7524,
       7623, 7722, 7821, 7920, 8019, 8118, 8217, 8316, 8415, 8514, 8613,
       8712, 8811, 8910, 9009, 9108, 9207, 9306, 9405, 9504, 9603, 9702,
       9801, 9900])

##### Leituras de temperatura 🌡️
A NOAA (Administração Nacional Oceânica e Atmosférica dos EUA) opera estações meteorológicas que medem as temperaturas da superfície em diferentes locais dos Estados Unidos. As leituras horárias estão [disponíveis publicamente](http://www.ncdc.noaa.gov/qclcd/QCLCD?prior=N).

Suponha que baixemos todos os dados de horários do site de San Diego, Califórnia, para o mês de dezembro de 2021. Para analisar os dados, queremos saber quando cada leitura foi feita, mas descobrimos que os dados não incluem os carimbos de data e hora das leituras (o momento em que cada uma foi feita).

No entanto, sabemos que a primeira leitura foi feita no primeiro instante de dezembro de 2021 (meia-noite do dia 1º de dezembro) e cada leitura subsequente foi feita exatamente 1 hora após a última.

**Pergunta 1.1.5.** Crie uma matriz do *tempo, em segundos, desde o início do mês* em que cada leitura horária foi feita. Nomeie-o como `registro_de_leituras`.

* **Dica 1:** Há 31 dias em dezembro, o que equivale a ($31 \times 24$) horas ou ($31 \times 24 \times 60 \times 60$) segundos.

* **Dica 2:** A função `len` também funciona em arrays. Certifique-se de que seu `registro_de_leituras` tenha $31 \times 24$ elementos, já que as leituras são feitas de hora em hora durante 31 dias.

In [11]:
hora = 60*60
segundosDezembro = 31*24*hora
registro_de_leituras = np.arange(0, segundosDezembro, hora)
print("Quantidade de registros: ", len(registro_de_leituras))
print("Momento em segundo dos registros: ", registro_de_leituras)

Quantidade de registros:  744
Momento em segundo dos registros:  [      0    3600    7200   10800   14400   18000   21600   25200   28800
   32400   36000   39600   43200   46800   50400   54000   57600   61200
   64800   68400   72000   75600   79200   82800   86400   90000   93600
   97200  100800  104400  108000  111600  115200  118800  122400  126000
  129600  133200  136800  140400  144000  147600  151200  154800  158400
  162000  165600  169200  172800  176400  180000  183600  187200  190800
  194400  198000  201600  205200  208800  212400  216000  219600  223200
  226800  230400  234000  237600  241200  244800  248400  252000  255600
  259200  262800  266400  270000  273600  277200  280800  284400  288000
  291600  295200  298800  302400  306000  309600  313200  316800  320400
  324000  327600  331200  334800  338400  342000  345600  349200  352800
  356400  360000  363600  367200  370800  374400  378000  381600  385200
  388800  392400  396000  399600  403200  406800  410400  4

## 1.2. Trabalhando com elementos únicos de arrays ("indexação")
Vamos trabalhar com um conjunto de dados mais interessante. A próxima célula cria uma matriz chamada `populacao` que inclui populações mundiais estimadas em cada ano de **1950** a **2022**. (As estimativas vêm da seguinte [base de dados internacional](https://www.census.gov/data-tools/demo/idb/#/country?COUNTRY_YEAR=2022&COUNTRY_YR_ANIM=2022), mantida pelo US Census Bureau.)

Em vez de digitar os dados manualmente, nós os carregamos de um arquivo em seu computador chamado `world_population_2022.csv`. Você aprenderá como ler dados de arquivos em breve.

In [12]:
# Não se preocupe sobre o que está acontecendo nessa célula por enquanto
populacao = pd.read_csv("data/world_population_2022.csv").get("Population").values
populacao

array([2557619597, 2594942227, 2636777090, 2682060684, 2730237675,
       2782111389, 2835315327, 2891368627, 2948159570, 3000742521,
       3043031253, 3084053711, 3140239653, 3210037409, 3281477826,
       3350773176, 3421097064, 3490825940, 3562887008, 3637819236,
       3713457589, 3791172327, 3867519813, 3943132388, 4017779234,
       4089387557, 4159536915, 4230430893, 4301282222, 4374940345,
       4445975606, 4527418598, 4610620221, 4694937687, 4777055423,
       4862317393, 4949951891, 5040273543, 5131575729, 5222662682,
       5315511894, 5403253915, 5490481497, 5568231516, 5650178207,
       5733211108, 5815333785, 5895837672, 5975189305, 6053955779,
       6132455985, 6211328357, 6290282107, 6369186797, 6448262425,
       6527056809, 6607396274, 6689442159, 6773319540, 6857160919,
       6939761510, 7022084781, 7105001721, 7188528811, 7271598780,
       7353476064, 7435151387, 7516769535, 7597066210, 7676686052,
       7756873419, 7831718605, 7905336896], dtype=int64)

Veja como obtemos o primeiro elemento de `populacao`, que é a população mundial no primeiro ano do conjunto de dados, 1950.

In [13]:
populacao[0]

2557619597

Observe que usamos colchetes aqui. Os colchetes sinalizam que estamos *acessando* um elemento do array. Colchetes em Python são como subscritos em matemática (igual $x_1$, $x_2$, ...).

O valor dessa expressão é o número 2557619597 (cerca de 2,5 bilhões), porque é a primeira coisa na matriz `populacao`.

Observe que escrevemos `populacao[0]`, não `populacao[1]`, para obter o primeiro elemento. Esta é uma convenção estranha na ciência da computação. 0 é chamado de *índice* do primeiro item. Seguindo essa lógica, então 3, por exemplo, é o índice do 4º item.

Aqui estão mais alguns exemplos. Nos exemplos, demos nomes às coisas que obtemos de `populacao`. Leia e execute cada célula.

In [14]:
# O terceiro elemento do array é a população em 1952.
populacao_1952 = populacao[2]
populacao_1952

2636777090

In [15]:
# O décimo terceiro elemento do array é a população em 1962 (que é 1950 + 12).
populacao_1962 = populacao[12]
populacao_1962

3140239653

In [16]:
# O 73º elemento do array é a população em 2022.
populacao_2022 = populacao[72]
populacao_2022

7905336896

In [17]:
# O array possui apenas 73 elementos, então isso não funciona.
#populacao_2023 = populacao[73]
#populacao_2023

# 🚨 Depois de executar esta célula, coloque um # antes de cada linha acima
# para garantir que ela não seja executada novamente.

**Questão 1.2.1.** Defina `populacao_1998` para a população mundial em 1998, obtendo o elemento apropriado de `populacao`.

In [18]:
ano = 1998-1950
populacao_1998 = populacao[ano]
populacao_1998

5975189305

## 1.3. Fazendo algo para cada elemento de um array
Arrays são úteis principalmente para realizar a mesma operação muitas vezes. Portanto, não precisamos acessar e trabalhar com frequência com elementos únicos.

##### Logaritmos
Aqui está uma pergunta simples que podemos fazer sobre a população mundial:

> Qual era o tamanho da população em *ordens de magnitude* em cada ano?

A função logaritmo é uma forma de medir o tamanho de um número. O logaritmo (base 10) de um número aumenta em 1 cada vez que multiplicamos o número por 10. É como uma medida de quantos dígitos decimais o número possui ou quão grande ele é em ordens de grandeza.

Poderíamos tentar responder nossa pergunta assim, usando a função `log10` do NumPy em cada elemento do array `populacao`:

In [19]:
magnitude_da_populacao_1950 = np.log10(populacao[0])
magnitude_da_populacao_1951 = np.log10(populacao[1])
magnitude_da_populacao_1952 = np.log10(populacao[2])
magnitude_da_populacao_1953 = np.log10(populacao[3])
magnitude_da_populacao_1951

9.414127693292606

Mas isso é tedioso e repetitivo. Deve haver uma maneira melhor!

Acontece que o `log10` do NumPy é bastante poderoso. Ele não apenas pode receber um único número (como `populacao[0]`) como entrada e retornar o logaritmo de um único número, mas **também** pode receber um array de números e retornar o logaritmo de cada elemento desse array!

Se você fornecer ao `log10` do NumPy um array como entrada, ele retornará um array do mesmo tamanho, onde o primeiro elemento do resultado é o logaritmo do primeiro elemento da entrada, o segundo elemento do resultado é o logaritmo de o segundo elemento da entrada e assim por diante.

<img src="https://github.com/ThiagoPoppe/monitoria_fcd2024/blob/main/labs/lab01/imagens/array_logarithm.jpg?raw=true">

Isso é chamado de aplicação *elementwise* (elemento a elemento) da função, uma vez que opera separadamente em cada elemento do array em que é chamada.

**Pergunta 1.3.1.** Use a função `log10` do NumPy para calcular os logaritmos da população mundial em cada ano. Dê ao resultado (uma matriz de 73 números) o nome `magnitude_da_populacao`. Seu código deve ser muito curto.

In [20]:
magnitude_da_populacao = np.log10(populacao)
magnitude_da_populacao

array([9.40783595, 9.41412769, 9.42107342, 9.4284686 , 9.43620046,
       9.44437451, 9.45260137, 9.46110346, 9.46955099, 9.47722873,
       9.48330641, 9.48912193, 9.49696279, 9.50651009, 9.51606947,
       9.52514503, 9.5341654 , 9.54292819, 9.55180205, 9.56084112,
       9.56977847, 9.57877353, 9.58743255, 9.59584136, 9.60398607,
       9.61165827, 9.61904498, 9.6263846 , 9.63359794, 9.64097214,
       9.64796708, 9.65585065, 9.66375935, 9.67162983, 9.67916028,
       9.6868433 , 9.69460098, 9.70245411, 9.71025074, 9.71789198,
       9.72554509, 9.73265538, 9.73961043, 9.74571728, 9.75206215,
       9.75839793, 9.76457465, 9.77054552, 9.77635167, 9.78203924,
       9.78763444, 9.79318449, 9.79867012, 9.80408399, 9.8094427 ,
       9.81471739, 9.82003035, 9.8253899 , 9.83080156, 9.83614434,
       9.84134455, 9.84646607, 9.85156419, 9.85664002, 9.86162991,
       9.86649268, 9.87128982, 9.87603123, 9.88064591, 9.88517378,
       9.8896867 , 9.89385707, 9.89792038])

##### Aritmética
A aritmética também funciona *elementwise* em arrays. Por exemplo, você pode dividir todos os números da população por 1 bilhão para obter números em bilhões:

In [21]:
populacao_em_bilhoes = populacao / 1000000000
populacao_em_bilhoes
#quantidade de pessoas em bilhoes:

array([2.5576196 , 2.59494223, 2.63677709, 2.68206068, 2.73023768,
       2.78211139, 2.83531533, 2.89136863, 2.94815957, 3.00074252,
       3.04303125, 3.08405371, 3.14023965, 3.21003741, 3.28147783,
       3.35077318, 3.42109706, 3.49082594, 3.56288701, 3.63781924,
       3.71345759, 3.79117233, 3.86751981, 3.94313239, 4.01777923,
       4.08938756, 4.15953692, 4.23043089, 4.30128222, 4.37494034,
       4.44597561, 4.5274186 , 4.61062022, 4.69493769, 4.77705542,
       4.86231739, 4.94995189, 5.04027354, 5.13157573, 5.22266268,
       5.31551189, 5.40325391, 5.4904815 , 5.56823152, 5.65017821,
       5.73321111, 5.81533378, 5.89583767, 5.9751893 , 6.05395578,
       6.13245598, 6.21132836, 6.29028211, 6.3691868 , 6.44826243,
       6.52705681, 6.60739627, 6.68944216, 6.77331954, 6.85716092,
       6.93976151, 7.02208478, 7.10500172, 7.18852881, 7.27159878,
       7.35347606, 7.43515139, 7.51676953, 7.59706621, 7.67668605,
       7.75687342, 7.8317186 , 7.9053369 ])

Você pode fazer o mesmo com adição, subtração, multiplicação e exponenciação (`**`). Por exemplo, você pode calcular uma gorjeta de vinte por cento em várias contas de restaurante de uma só vez:

In [22]:
contas_dos_restaurantes = np.array([20.12, 39.90, 31.01])
print("Conta dos restaurantes:\t", contas_dos_restaurantes)

gorjetas = 0.2 * contas_dos_restaurantes
print("Gorjetas:\t\t", gorjetas)

total = contas_dos_restaurantes + gorjetas
print("Total:\t\t\t", total)

Conta dos restaurantes:	 [20.12 39.9  31.01]
Gorjetas:		 [4.024 7.98  6.202]
Total:			 [24.144 47.88  37.212]


<img src="https://github.com/ThiagoPoppe/monitoria_fcd2024/blob/main/labs/lab01/imagens/array_multiplication.jpg?raw=true">

**Pergunta 1.3.2.** Suponha que a cobrança total em um restaurante seja a conta original mais a gorjeta (20%). Isso significa que podemos multiplicar a fatura original por 1.2 para obter a cobrança total. Calcule a cobrança total de cada conta em `conta_dos_restaurantes` e dê ao array resultante o nome de `cobrancas_totais`.

In [23]:
cobrancas_totais = contas_dos_restaurantes * 1.2
cobrancas_totais

array([24.144, 47.88 , 37.212])

Vamos ler alguns dados para usar na próxima pergunta.

In [24]:
mais_contas_de_restaurantes = pd.read_csv("data/more_restaurant_bills.csv").get("Bill").values

**Pergunta 1.3.3.** O array `mais_contas_de_restaurantes` contém 100.000 contas! Calcule a cobrança total de cada uma, assumindo novamente uma gorjeta de vinte por cento, e dê ao array resultante o nome `mais_cobrancas_totais`.

In [25]:
mais_cobrancas_totais = mais_contas_de_restaurantes * 1.2
mais_cobrancas_totais

array([20.244, 20.892, 12.216, ..., 19.308, 18.336, 35.664])

A função `np.sum` leva um único array de números como argumento. Ele retorna a soma de todos os números desse array (portanto, retorna um único número, não um array).

**Pergunta 1.3.4.** Qual foi a soma de todas as contas em `mais_contas_de_restaurantes`, **incluindo gorjetas**?

In [26]:
soma_das_contas = np.sum(mais_cobrancas_totais)
soma_das_contas

1795730.0639999998

##### Potências de Dois
As potências de 2 ($2^0 = 1$, $2^1 = 2$, $2^2 = 4$, etc) surgem frequentemente na ciência da computação. (Por exemplo, você deve ter notado que o armazenamento em smartphones ou computadores vem em potências de 2, como 64 GB, 128 GB ou 256 GB.)

**Pergunta 1.3.5.** Use `np.arange` e o operador de exponenciação `**` para criar um array contendo as primeiras 40 potências de 2, começando em $2^0=1$.

* **Dica 1:** Seu kernel “morreu” quando você executou sua solução? Existe uma resposta incorreta comum para esse problema que tenta criar um array com tantas entradas que o Python desiste e trava. Se isso acontecer com você, verifique sua resposta!

* **Dica 2:** Talvez comece com as primeiras 5 potências de dois. Depois de fazer isso funcionar, tente todos os 40. Em nenhum momento você deve escrever manualmente `0, 1, 2, 3, 4, ...`; se você estiver tentando isso, leia novamente algumas células anteriores desse notebook.

In [27]:
#potencia = 2**39

potencias_de_2 = 2**(np.arange(39))
potencias_de_2

array([          1,           2,           4,           8,          16,
                32,          64,         128,         256,         512,
              1024,        2048,        4096,        8192,       16384,
             32768,       65536,      131072,      262144,      524288,
           1048576,     2097152,     4194304,     8388608,    16777216,
          33554432,    67108864,   134217728,   268435456,   536870912,
        1073741824, -2147483648,           0,           0,           0,
                 0,           0,           0,           0], dtype=int32)

# 2. DataFrames

## 2.1. Introdução

Para uma coleção de coisas no mundo, um array é útil para descrever um único atributo de cada coisa. Por exemplo, para a coleção de estados dos EUA, uma matriz poderia descrever a área territorial de cada um. As tabelas ampliam essa ideia descrevendo vários atributos para cada elemento de uma coleção. Numa tabela de estados, por exemplo, podemos registar a área territorial, a população, a capital do estado e o nome do governador. Em outras palavras, as tabelas rastreiam muitas entidades (indivíduos, armazenados como linhas) e, para cada entidade, muitos atributos (recursos, armazenados como colunas).

Na célula abaixo temos dois arrays. O primeiro contém a população mundial em cada ano (conforme estimado pelo US Census Bureau), e o segundo contém os próprios anos (em ordem, de modo que os primeiros elementos na população e as matrizes de anos correspondam).

In [28]:
anos = np.arange(1950, 2022+1)
quantidade_populacional = pd.read_csv("data/world_population_2022.csv").get("Population").values

print("Coluna de população:", quantidade_populacional)
print("Coluna dos anos:", anos)

Coluna de população: [2557619597 2594942227 2636777090 2682060684 2730237675 2782111389
 2835315327 2891368627 2948159570 3000742521 3043031253 3084053711
 3140239653 3210037409 3281477826 3350773176 3421097064 3490825940
 3562887008 3637819236 3713457589 3791172327 3867519813 3943132388
 4017779234 4089387557 4159536915 4230430893 4301282222 4374940345
 4445975606 4527418598 4610620221 4694937687 4777055423 4862317393
 4949951891 5040273543 5131575729 5222662682 5315511894 5403253915
 5490481497 5568231516 5650178207 5733211108 5815333785 5895837672
 5975189305 6053955779 6132455985 6211328357 6290282107 6369186797
 6448262425 6527056809 6607396274 6689442159 6773319540 6857160919
 6939761510 7022084781 7105001721 7188528811 7271598780 7353476064
 7435151387 7516769535 7597066210 7676686052 7756873419 7831718605
 7905336896]
Coluna dos anos: [1950 1951 1952 1953 1954 1955 1956 1957 1958 1959 1960 1961 1962 1963
 1964 1965 1966 1967 1968 1969 1970 1971 1972 1973 1974 1975 1976 1977
 19

In [29]:
quantidade_populacional[quantidade_populacional>7000000000].min()


7022084781

Suponha que queiramos responder a esta pergunta:

> Quando é que a população mundial ultrapassou os 7 mil milhões?

Você poderia tecnicamente responder a essa pergunta apenas olhando para as matrizes, mas é um pouco complicado, pois seria necessário contar a posição onde a população ultrapassou pela primeira vez os 7 bilhões e, em seguida, encontrar o elemento correspondente na matriz de anos. Em casos como estes, pode ser mais fácil colocar os dados em uma tabela.

Assim como `numpy` fornece arrays, um pacote popular chamado `pandas` fornece **DataFrames**, que é o nome em `pandas` para **tabelas**. `pandas` é *a* ferramenta para fazer ciência de dados em Python. Infelizmente, `pandas` não é tão fofo quanto o nome sugere: é muito complicado e pode ser um tanto difícil de aprender.

Em vez de usar `pandas`, usaremos um pacote que criamos especificamente para o curso do DSC10. É um *subconjunto* de `pandas`, incluindo apenas as partes que os criadores acharam necessárias e descartando todo o resto . Por ser menor (e mais fofo), eles deram para esse módulo o nome carinhoso de `babypandas`.

<img src='https://github.com/ThiagoPoppe/monitoria_fcd2024/blob/main/labs/lab01/imagens/pandas-babypandas.jpg?raw=true' width=400>

Você pode importar `babypandas` usando o seguinte código:

In [30]:
import pandas as pd


O bom de `babypandas` é que é mais fácil de aprender *mas* cada pedaço de código que você escreve usando `babypandas` funcionará com `pandas` também. Se você é especialista em ciência de dados ou apenas fará muitas análises de dados em Python, verá muitos `pandas` no futuro.

A célula abaixo:

- cria um DataFrame vazio usando a expressão `pd.DataFrame()`,
- atribui duas colunas ao DataFrame chamando `assign`,
- atribui o DataFrame resultante ao nome `populacao_df` e, finalmente
- exibe `populacao_df` para que possamos ver o DataFrame que criamos.

`"Populacao"` e `"Ano"` são rótulos de coluna que escolhemos. Poderíamos ter escolhido qualquer coisa, mas é uma boa ideia escolher nomes que sejam descritivos e não muito longos.

In [31]:
populacao_df = pd.DataFrame().assign(
    Populacao=quantidade_populacional,
    Ano=anos
)

#populacao_df.set_index('Ano')
populacao_df

Populacao   Ano
0   2557619597  1950
1   2594942227  1951
2   2636777090  1952
3   2682060684  1953
4   2730237675  1954
..         ...   ...
68  7597066210  2018
69  7676686052  2019
70  7756873419  2020
71  7831718605  2021
72  7905336896  2022

[73 rows x 2 columns]

Agora os dados estão todos juntos em um único DataFrame! É muito mais fácil analisar esses dados. Se você precisa saber qual era a população em 2011, por exemplo, você pode saber com um simples olhar. Revisitaremos este DataFrame mais tarde.

In [32]:
populacao_df[populacao_df['Populacao'] > 7000000000]["Ano"].min()

2011

**Questão 2.1.1.** Na célula abaixo, criamos 2 arrays. Usando as etapas acima, atribua `top_10_filmes` a um DataFrame que possui duas colunas chamadas `Avaliacao` e `Nome`, que contêm `top_10_avaliacoes` e `top_10_nomes` respectivamente.

In [33]:
top_10_avaliacoes = np.array([9.2, 9.2, 9., 8.9, 8.9, 8.9, 8.9, 8.9, 8.9, 8.8])
top_10_nomes = np.array([
        'The Shawshank Redemption (1994)',
        'The Godfather (1972)',
        'The Godfather: Part II (1974)',
        'Pulp Fiction (1994)',
        "Schindler's List (1993)",
        'The Lord of the Rings: The Return of the King (2003)',
        '12 Angry Men (1957)',
        'The Dark Knight (2008)',
        'Il buono, il brutto, il cattivo (1966)',
        'The Lord of the Rings: The Fellowship of the Ring (2001)'
])

In [34]:
top_10_filmes = pd.DataFrame().assign(
    nomes = top_10_nomes,
    avaliacoes = top_10_avaliacoes
)
top_10_filmes.sort_values(by = 'avaliacoes', ascending=False)

nomes  avaliacoes
0                    The Shawshank Redemption (1994)         9.2
1                               The Godfather (1972)         9.2
2                      The Godfather: Part II (1974)         9.0
3                                Pulp Fiction (1994)         8.9
4                            Schindler's List (1993)         8.9
5  The Lord of the Rings: The Return of the King ...         8.9
6                                12 Angry Men (1957)         8.9
7                             The Dark Knight (2008)         8.9
8             Il buono, il brutto, il cattivo (1966)         8.9
9  The Lord of the Rings: The Fellowship of the R...         8.8

Suponha que você queira adicionar suas próprias classificações a este DataFrame. A célula abaixo contém sua avaliação de cada filme:

In [35]:
minhas_avaliacoes = [8, 2, 1, 9, 7, 10, 6, 4, 3, 5]

**Pergunta 2.1.2** Você também pode usar o método `assign` para adicionar uma coluna a um DataFrame já existente. Crie um novo DataFrame chamado `com_minhas_avaliacoes` adicionando uma coluna chamada `MinhaAvaliacao` ao DataFrame em `top_10_filmes`.

In [36]:
com_minhas_avaliacoes = pd.DataFrame().assign(
    nomes = top_10_nomes,
    avaliacao_geral = top_10_avaliacoes,
    avaliacao_pessoal = minhas_avaliacoes
)
com_minhas_avaliacoes

nomes  avaliacao_geral  \
0                    The Shawshank Redemption (1994)              9.2   
1                               The Godfather (1972)              9.2   
2                      The Godfather: Part II (1974)              9.0   
3                                Pulp Fiction (1994)              8.9   
4                            Schindler's List (1993)              8.9   
5  The Lord of the Rings: The Return of the King ...              8.9   
6                                12 Angry Men (1957)              8.9   
7                             The Dark Knight (2008)              8.9   
8             Il buono, il brutto, il cattivo (1966)              8.9   
9  The Lord of the Rings: The Fellowship of the R...              8.8   

   avaliacao_pessoal  
0                  8  
1                  2  
2                  1  
3                  9  
4                  7  
5                 10  
6                  6  
7                  4  
8                  3  
9                  5

## 2.2. Índices

Você deve ter notado que o DataFrame de populações contém o que parece ser uma coluna extra e sem rótulo à esquerda com os números de 0 a 65. **Isto não é uma coluna, é o que chamamos de *índice***. O índice contém os rótulos das linhas. Enquanto as colunas deste DataFrame são rotuladas como `"Populacao"` e `"Ano"`, as linhas são rotuladas como 0, 1, ..., 65.

Por padrão, `babypandas` não sabe como rotular as linhas, então apenas as numera (começando com 0). É claro que, neste caso, faz mais sentido usar o ano como rótulo da linha. Podemos fazer isso dizendo ao `babypandas` para definir a coluna `"Ano"` como índice:

In [37]:
populacao_por_ano = populacao_df.set_index('Ano')
populacao_por_ano

Populacao
Ano             
1950  2557619597
1951  2594942227
1952  2636777090
1953  2682060684
1954  2730237675
...          ...
2018  7597066210
2019  7676686052
2020  7756873419
2021  7831718605
2022  7905336896

[73 rows x 1 columns]

Como veremos, isso faz mais do que deixar o DataFrame mais bonito – é muito útil também.

**Pergunta 2.2.1** Crie um novo DataFrame chamado `top_10_filmes_por_nome` pegando o DataFrame que você criou acima, `top_10_filmes`, e definindo o índice como a coluna `Nome`.

In [38]:
top_10_filmes_por_nome = top_10_filmes.set_index('nomes')
top_10_filmes_por_nome

avaliacoes
nomes                                                         
The Shawshank Redemption (1994)                            9.2
The Godfather (1972)                                       9.2
The Godfather: Part II (1974)                              9.0
Pulp Fiction (1994)                                        8.9
Schindler's List (1993)                                    8.9
The Lord of the Rings: The Return of the King (...         8.9
12 Angry Men (1957)                                        8.9
The Dark Knight (2008)                                     8.9
Il buono, il brutto, il cattivo (1966)                     8.9
The Lord of the Rings: The Fellowship of the Ri...         8.8

Você pode obter um array de nomes de linhas usando `.index`. Por exemplo, o array de nomes de linhas do DataFrame `populacao_por_ano` é:

In [39]:
populacao_por_ano.index

Index([1950, 1951, 1952, 1953, 1954, 1955, 1956, 1957, 1958, 1959, 1960, 1961,
       1962, 1963, 1964, 1965, 1966, 1967, 1968, 1969, 1970, 1971, 1972, 1973,
       1974, 1975, 1976, 1977, 1978, 1979, 1980, 1981, 1982, 1983, 1984, 1985,
       1986, 1987, 1988, 1989, 1990, 1991, 1992, 1993, 1994, 1995, 1996, 1997,
       1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009,
       2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021,
       2022],
      dtype='int32', name='Ano')

**Questão 2.2.2** Usando o código acima, atribua a `decimo_filme` o nome do décimo filme em `top_10_filmes_por_nome`.

* **Dica:** Lembre-se de que o índice é um array e usamos colchetes para acessar os elementos de um array. E também que o índice 0 corresponde ao 1º elemento do array.

In [40]:
decimo_filme = top_10_filmes_por_nome.index[10-1]
decimo_filme

'The Lord of the Rings: The Fellowship of the Ring (2001)'

## 2.3 Lendo um DataFrame de um arquivo
Na maioria dos casos, não teremos o trabalho de digitar todos os dados manualmente. Em vez disso, podemos usar funções fornecidas por `babypandas` para ler dados de arquivos externos.

A função `pd.read_csv()` pega um argumento, um caminho para um arquivo de dados (uma string) e retorna um DataFrame. Existem muitos formatos de arquivos de dados, mas CSV (*comma separated values*, ou "valores separados por vírgula") é o mais comum.

**Questão 2.3.1.** O arquivo `data/imdb.csv` contém informações sobre os 250 filmes mais bem avaliados no IMDb. Carregue-o como um DataFrame chamado `imdb`.

In [41]:
imdb = pd.read_csv("data/imdb.csv")
imdb

Votes  Rating                                  Title  Year  Decade
0      88355     8.4                                      M  1931    1930
1     132823     8.3                    Singin' in the Rain  1952    1950
2      74178     8.3                          All About Eve  1950    1950
3     635139     8.6                                   Léon  1994    1990
4     145514     8.2                       The Elephant Man  1980    1980
..       ...     ...                                    ...   ...     ...
245  1078416     8.7                           Forrest Gump  1994    1990
246    31003     8.1                  Le salaire de la peur  1953    1950
247   167076     8.2                               3 Idiots  2009    2000
248    91689     8.1                                Network  1976    1970
249   589477     8.3  Eternal Sunshine of the Spotless Mind  2004    2000

[250 rows x 5 columns]

Observe os `...` no meio do DataFrame. Isso significa que muitas linhas foram omitidas. Este DataFrame é grande o suficiente para que apenas algumas de suas linhas sejam exibidas, mas as outras ainda estão lá. São 250 filmes no total.

De onde veio o `imdb.csv`? Se você entrar no diretório `data/`, deverá ver um arquivo chamado `imdb.csv`.

Abra o arquivo `imdb.csv` nessa pasta e observe o formato. O que você percebe? A terminação do nome do arquivo `.csv` indica que este arquivo está no formato [CSV (comma-separated value)](http://edoceo.com/utilitas/csv-file-format).

**Pergunta 2.3.2.** Este é um conjunto de dados de filmes, portanto faz sentido usar o título do filme como rótulo da linha. Crie um novo DataFrame chamado `imdb_por_nome` que usa o título do filme como índice.

In [42]:
imdb_por_nome = imdb.set_index("Title")
imdb_por_nome

Votes  Rating  Year  Decade
Title                                                               
M                                        88355     8.4  1931    1930
Singin' in the Rain                     132823     8.3  1952    1950
All About Eve                            74178     8.3  1950    1950
Léon                                    635139     8.6  1994    1990
The Elephant Man                        145514     8.2  1980    1980
...                                        ...     ...   ...     ...
Forrest Gump                           1078416     8.7  1994    1990
Le salaire de la peur                    31003     8.1  1953    1950
3 Idiots                                167076     8.2  2009    2000
Network                                  91689     8.1  1976    1970
Eternal Sunshine of the Spotless Mind   589477     8.3  2004    2000

[250 rows x 4 columns]

## 2.4. Series



Suponha que estejamos interessados ​​principalmente nas classificações de filmes. Para extrair apenas esta coluna do DataFrame, usamos o método `.get`:

In [43]:
avaliacoes = imdb_por_nome.get('Rating')
avaliacoes

Title
M                                        8.4
Singin' in the Rain                      8.3
All About Eve                            8.3
Léon                                     8.6
The Elephant Man                         8.2
                                        ... 
Forrest Gump                             8.7
Le salaire de la peur                    8.1
3 Idiots                                 8.2
Network                                  8.1
Eternal Sunshine of the Spotless Mind    8.3
Name: Rating, Length: 250, dtype: float64

Observe como não apenas as classificações do filme foram retornadas, mas também o nome do filme! Isto ocorre precisamente porque definimos o título do filme como o índice! Por exemplo, se tivéssemos solicitado a coluna `"Rating"` do DataFrame original, `imdb`, veríamos:

In [44]:
imdb.get('Rating')

0      8.4
1      8.3
2      8.3
3      8.6
4      8.2
      ... 
245    8.7
246    8.1
247    8.2
248    8.1
249    8.3
Name: Rating, Length: 250, dtype: float64

Esta é uma forma pela qual os índices são muito úteis – eles fornecem rótulos significativos para os dados.

À primeira vista, pode parecer que pedir uma coluna usando `.get` retorna um DataFrame com uma coluna, mas isso não está certo. Em vez disso, ele retorna um tipo especial de coisa chamado *Series*:

In [45]:
type(imdb_por_nome.get('Rating'))

pandas.core.series.Series

Você pode pensar em uma `Series` como um array com um índice. Enquanto as matrizes são sequências simples de números sem rótulos, `Series` pode ter rótulos. Isso geralmente é muito útil.

`avaliacoes` agora é uma `Series` que contém a coluna de classificações de filmes. Suponha que estejamos interessados ​​na avaliação de um filme específico: _Alien_. Para fazer isso, usaremos o "*acessador*" `.loc` que extrai um valor da Série em um *local* específico:

In [46]:
avaliacoes.loc["Alien"]

8.5

Há algumas coisas a serem observadas aqui. Primeiro, esses são colchetes em torno de `"Alien"`. Isso ocorre porque `.loc` não é um método, mas um "*acessador*". Os colchetes sinalizam que iremos extrair um elemento da `Série`. Segundo, passamos o rótulo como uma string.

**Pergunta 2.4.1.** Encontre a avaliação de _3 Idiotas_ (*3 Idiots*).

In [47]:
avaliacao_de_tres_idiotas = avaliacoes.loc["3 Idiots"]
avaliacao_de_tres_idiotas

8.2

Agora suponha que quiséssemos saber o ano em que _Alien_ foi lançado. Poderíamos fazer isso obtendo primeiro a coluna dos anos:

In [48]:
anos = imdb_por_nome.get('Year')
anos

Title
M                                        1931
Singin' in the Rain                      1952
All About Eve                            1950
Léon                                     1994
The Elephant Man                         1980
                                         ... 
Forrest Gump                             1994
Le salaire de la peur                    1953
3 Idiots                                 2009
Network                                  1976
Eternal Sunshine of the Spotless Mind    2004
Name: Year, Length: 250, dtype: int64

E então usando `.loc` para obter a entrada correta:

In [49]:
anos.loc['Alien']

1979

Também poderíamos fazer isso em uma única etapa *encadeando* as operações:

In [50]:
imdb_por_nome.get('Year').loc['Alien']

1979

Isso funciona porque o Python primeiro avalia `imdb_por_nome.get('Year')` como uma `Series`. Em seguida, avalia `.loc['Alien']` para retornar o ano.

O encadeamento é usado com bastante frequência e pode ser útil. Apenas certifique-se de não encadear muitas coisas que tornem seu código difícil de ler. Você sempre pode salvar um resultado intermediário em uma variável.

**Pergunta 2.4.2** Encontre a década em que _Gone Girl_ foi lançado usando encadeamento.

*Dica*: `imbd_por_nome` possui uma coluna chamada `"Decade"` (*Década*).

In [51]:
decada = imdb_por_nome.get('Decade')
decada

Title
M                                        1930
Singin' in the Rain                      1950
All About Eve                            1950
Léon                                     1990
The Elephant Man                         1980
                                         ... 
Forrest Gump                             1990
Le salaire de la peur                    1950
3 Idiots                                 2000
Network                                  1970
Eternal Sunshine of the Spotless Mind    2000
Name: Decade, Length: 250, dtype: int64

In [52]:
lancamento_Gone_Girl = decada.loc['Gone Girl']
lancamento_Gone_Girl

2010

# 3. Analisando conjuntos de dados

Com apenas alguns métodos DataFrame, podemos responder algumas questões interessantes sobre o conjunto de dados IMDb.

Se quisermos apenas as avaliações dos filmes, podemos usar `.get`:

In [53]:
avaliacoes = imdb_por_nome.get("Rating")
avaliacoes

Title
M                                        8.4
Singin' in the Rain                      8.3
All About Eve                            8.3
Léon                                     8.6
The Elephant Man                         8.2
                                        ... 
Forrest Gump                             8.7
Le salaire de la peur                    8.1
3 Idiots                                 8.2
Network                                  8.1
Eternal Sunshine of the Spotless Mind    8.3
Name: Rating, Length: 250, dtype: float64

Lembre-se de que `avaliacoes` é uma série. Objetos de série possuem alguns métodos úteis.

**Pergunta 3.1.** Encontre a maior avaliação no conjunto de dados.

*Dica:* Digite `avaliacoes.` e pressione Tab para ver uma lista dos métodos disponíveis. Existe algum que parece útil?

In [54]:
maior_avaliacao = avaliacoes.max()
maior_avaliacao

9.2

Você provavelmente quer saber o *nome* do filme de maior avaliação que encontrou anteriormente! Para fazer isso, podemos ordenar toda a série usando o método `.sort_values`:

In [55]:
avaliacoes.sort_values()

Title
Akira                               8.0
Per un pugno di dollari             8.0
Guardians of the Galaxy             8.0
The Man Who Shot Liberty Valance    8.0
Underground                         8.0
                                   ... 
Schindler's List                    8.9
12 Angry Men                        8.9
The Godfather: Part II              9.0
The Shawshank Redemption            9.2
The Godfather                       9.2
Name: Rating, Length: 250, dtype: float64

Portanto, na verdade, existem dois filmes com maior audiência no conjunto de dados: *Um Sonho de Liberdade* e *O Poderoso Chefão*.

Observe que estamos ordenando pelas avaliações, e não pelos rótulos! Além disso, o rótulo segue a ordenação conforme a sua avaliação. Isto é exatamente o que queremos.

Quando utilizamos o método `sort_values`, a `Series` resultante tem os dados ordenados em ordem crescente, do menor ao maior. Este é o comportamento padrão de `sort_values`, mas podemos mudar isso. Se quiséssemos os filmes com melhor avaliação no topo, precisaríamos especificar que a avaliação não deveria ser em ordem crescente com um *argumento keyword* ("palavra-chave") opcional:


In [56]:
avaliacoes.sort_values(ascending=False)

Title
The Godfather                             9.2
The Shawshank Redemption                  9.2
The Godfather: Part II                    9.0
12 Angry Men                              8.9
Il buono, il brutto, il cattivo (1966)    8.9
                                         ... 
Monsters, Inc. (2001)                     8.0
The Big Sleep                             8.0
X-Men: Days of Future Past                8.0
Roman Holiday                             8.0
Kumonosu-jô                               8.0
Name: Rating, Length: 250, dtype: float64

Se definirmos o argumento `ascending` como `True`, obteremos o mesmo resultado como se não o definissemos. Isso é o que queremos dizer quando dizemos que o comportamento padrão de `sort_values` é classificar em ordem crescente. Confirme se as próximas duas células fornecem a mesma saída.

In [57]:
avaliacoes.sort_values(ascending=True)

Title
Akira                               8.0
Per un pugno di dollari             8.0
Guardians of the Galaxy             8.0
The Man Who Shot Liberty Valance    8.0
Underground                         8.0
                                   ... 
Schindler's List                    8.9
12 Angry Men                        8.9
The Godfather: Part II              9.0
The Shawshank Redemption            9.2
The Godfather                       9.2
Name: Rating, Length: 250, dtype: float64

In [58]:
avaliacoes.sort_values()

Title
Akira                               8.0
Per un pugno di dollari             8.0
Guardians of the Galaxy             8.0
The Man Who Shot Liberty Valance    8.0
Underground                         8.0
                                   ... 
Schindler's List                    8.9
12 Angry Men                        8.9
The Godfather: Part II              9.0
The Shawshank Redemption            9.2
The Godfather                       9.2
Name: Rating, Length: 250, dtype: float64

Não só podemos ordenar séries, mas também ordenar DataFrames inteiros. Quando fazemos isso, temos que especificar a coluna pela qual iremos ordenar:

In [59]:
imdb_por_nome.sort_values('Rating')

Votes  Rating  Year  Decade
Title                                                          
Akira                               91652     8.0  1988    1980
Per un pugno di dollari            124671     8.0  1964    1960
Guardians of the Galaxy            527349     8.0  2014    2010
The Man Who Shot Liberty Valance    49135     8.0  1962    1960
Underground                         39447     8.0  1995    1990
...                                   ...     ...   ...     ...
Schindler's List                   761224     8.9  1993    1990
12 Angry Men                       384187     8.9  1957    1950
The Godfather: Part II             692753     9.0  1974    1970
The Shawshank Redemption          1498733     9.2  1994    1990
The Godfather                     1027398     9.2  1972    1970

[250 rows x 4 columns]

Da mesma forma, podemos especificar que a ordenação deve estar em ordem decrescente:

In [60]:
imdb_por_nome.sort_values('Rating', ascending=False)

Votes  Rating  Year  Decade
Title                                                                
The Godfather                           1027398     9.2  1972    1970
The Shawshank Redemption                1498733     9.2  1994    1990
The Godfather: Part II                   692753     9.0  1974    1970
12 Angry Men                             384187     8.9  1957    1950
Il buono, il brutto, il cattivo (1966)   447875     8.9  1966    1960
...                                         ...     ...   ...     ...
Monsters, Inc. (2001)                    500576     8.0  2001    2000
The Big Sleep                             59578     8.0  1946    1940
X-Men: Days of Future Past               427099     8.0  2014    2010
Roman Holiday                             87437     8.0  1953    1950
Kumonosu-jô                               26012     8.0  1957    1950

[250 rows x 4 columns]

Alguns detalhes sobre a ordenação de um DataFrame:

1. O primeiro argumento para `sort_values` é o nome de uma coluna pela qual iremos ordenar.
2. Se a coluna contiver strings, `sort` ordenará em ordem alfabética; se a coluna tiver números, ela será ordenada numericamente.
3. `imdb_por_nome.sort_values("Rating")` retorna um novo DataFrame; o DataFrame `imdb_por_nome` não é modificado. Por exemplo, se chamarmos `imdb_por_nome.sort("Rating")`, então executar `imdb_por_nome` por si só ainda retornaria o DataFrame não ordenado. Para salvar o resultado, você deve atribuí-lo a uma nova variável.
4. As linhas sempre permanecem juntas quando um DataFrame é ordenado. Não faria sentido ordenar apenas uma coluna e deixar as outras colunas em paz. Por exemplo, neste caso, se ordenássemos apenas a coluna `"Rating"`, todos os filmes terminariam com avaliações erradas.

**Pergunta 3.2.** Crie uma versão de `imdb_por_nome` que seja ordenada cronologicamente, com os filmes mais antigos primeiro. Chame-o de `imdb_ordenado`.

In [61]:
imdb_ordenado = imdb_por_nome.sort_values("Year", ascending=True)
imdb_ordenado

Votes  Rating  Year  Decade
Title                                                 
The Kid                    55784     8.3  1921    1920
The Gold Rush              58506     8.2  1925    1920
The General                46332     8.2  1926    1920
Metropolis                 98794     8.3  1927    1920
M                          88355     8.4  1931    1930
...                          ...     ...   ...     ...
The Grand Budapest Hotel  369141     8.1  2014    2010
Relatos salvajes           46987     8.0  2014    2010
Interstellar              689541     8.6  2014    2010
Mad Max: Fury Road        262425     8.3  2015    2010
Inside Out (2015/I)        79615     8.5  2015    2010

[250 rows x 4 columns]

**Pergunta 3.3.** Qual é o título do filme mais antigo no conjunto de dados? Você poderia simplesmente procurar isso na saída da célula anterior. Em vez disso, escreva o código Python para descobrir.

* **Dica:** Lembre-se de que o índice é um array.

In [62]:
#da pra fazer essa questao bem mais curta. (OBS: Fazer depois)
titulo_do_filme_mais_antigo = imdb.set_index('Year').sort_values('Year', ascending=True).loc[imdb["Year"].min()].get("Title")
titulo_do_filme_mais_antigo

#imdb_ordenado.get('Title').iloc[5-1]   -> Nao eh possivel fazer isso pois o title eh o index desse df.

'The Kid'

Suponha que queiramos obter a avaliação do filme mais antigo no DataFrame. Uma maneira de fazer isso é primeiro encontrar o rótulo de índice do filme mais antigo (o que já fizemos). Em seguida, extraímos a coluna `"Rating"` e usamos `.loc` para encontrar a avaliação do filme mais antigo.

In [63]:
imdb_ordenado.get('Rating').loc[titulo_do_filme_mais_antigo]

8.3

Porém, existe uma maneira mais rápida. Uma série não possui apenas um acessador `.loc`, mas também um acessador `.iloc`. Enquanto `.loc` procura coisas por *rótulo*, `.iloc` procura elementos por *posição inteira*.

Vamos lembrar o que está na coluna `"Rating"`:

In [64]:
imdb_ordenado.get('Rating')

Title
The Kid                     8.3
The Gold Rush               8.2
The General                 8.2
Metropolis                  8.3
M                           8.4
                           ... 
The Grand Budapest Hotel    8.1
Relatos salvajes            8.0
Interstellar                8.6
Mad Max: Fury Road          8.3
Inside Out (2015/I)         8.5
Name: Rating, Length: 250, dtype: float64

Se quisermos a avaliação da primeira linha, podemos usar `.iloc[0]`:

In [65]:
imdb_ordenado.get('Rating').iloc[0]

8.3

Isso retorna exatamente a mesma coisa que `imdb_ordenado.get('Rating').loc['The Kid']`; essas são duas maneiras de fazer a mesma coisa. Normalmente é mais conveniente acessar um elemento por seu rótulo do que por sua posição inteira, mas é bom saber `.loc` e `.iloc`.

**Pergunta 3.4.** Qual é a avaliação do quinto filme mais antigo no conjunto de dados? Você poderia simplesmente procurar isso na saída da célula anterior. Em vez disso, escreva o código Python para descobrir.

In [66]:
avaliacao_do_quinto_filme_mais_antigo = imdb_ordenado.get('Rating').iloc[5-1]
avaliacao_do_quinto_filme_mais_antigo

8.4

# 4. Encontrar partes de um conjunto de dados

Suponha que você esteja interessado em filmes da década de 1950. Ordenar o DataFrame por ano não ajuda, porque a década de 1950 está no meio do conjunto de dados. Em vez disso, usaremos um recurso de Series que nos permite comparar facilmente cada elemento em uma coluna com um valor específico.

Primeiro lembre-se que podemos usar `.get` para extrair uma única coluna. O resultado não é um DataFrame, mas sim uma Series:

In [67]:
imdb_por_nome.get('Decade')

Title
M                                        1930
Singin' in the Rain                      1950
All About Eve                            1950
Léon                                     1990
The Elephant Man                         1980
                                         ... 
Forrest Gump                             1990
Le salaire de la peur                    1950
3 Idiots                                 2000
Network                                  1970
Eternal Sunshine of the Spotless Mind    2000
Name: Decade, Length: 250, dtype: int64

Queremos verificar se cada filme foi lançado na década de 1950. Python nos dá uma maneira de verificar se duas coisas são iguais com `==` (lembre-se que `=` já está sendo usado para outro propósito: atribui valores a variáveis nomes):

In [68]:
3 == 4

False

In [69]:
3 == 3

True

`True` e `False` são instâncias de um tipo que não vimos antes:

In [70]:
type(True)

bool

`bool` significa "Boolean", em homenagem ao lógico inglês [George Boole](https://en.wikipedia.org/wiki/George_Boole). Dizemos que "True" e "False" são valores *Booleanos*.

Acontece que podemos facilmente verificar se *cada* um dos elementos em uma `Series` é igual a alguma coisa:

In [71]:
imdb_por_nome.get('Decade') == 1950

Title
M                                        False
Singin' in the Rain                       True
All About Eve                             True
Léon                                     False
The Elephant Man                         False
                                         ...  
Forrest Gump                             False
Le salaire de la peur                     True
3 Idiots                                 False
Network                                  False
Eternal Sunshine of the Spotless Mind    False
Name: Decade, Length: 250, dtype: bool

Vemos que o resultado é uma nova série que tem `Verdadeiro` apenas onde a década foi 1950, e `Falso` em todos os outros lugares. Dizemos que a Series resultante é uma Series de *Booleanos*, ou uma *Series Booleana*.

Vamos chamar esse resultado de `eh_de_1950s`. Seu nome pode ser lido como se fosse uma pergunta: “esse filme é da década de 1950”?

In [72]:
eh_de_1950s = imdb_por_nome.get('Decade') == 1950
eh_de_1950s

Title
M                                        False
Singin' in the Rain                       True
All About Eve                             True
Léon                                     False
The Elephant Man                         False
                                         ...  
Forrest Gump                             False
Le salaire de la peur                     True
3 Idiots                                 False
Network                                  False
Eternal Sunshine of the Spotless Mind    False
Name: Decade, Length: 250, dtype: bool

Cada linha é uma resposta a esta pergunta. *O Homem Elefante* é da década de 1950? `Falso`. *Tudo sobre Eva* é da década de 1950? `Verdadeiro`.

Podemos usar `eh_de_1950s` para selecionar apenas as linhas de `imdb_por_nome` para as quais a resposta é `Verdadeiro`. A sintaxe para isso é:

In [73]:
imdb_por_nome[eh_de_1950s]

Votes  Rating  Year  Decade
Title                                                     
Singin' in the Rain           132823     8.3  1952    1950
All About Eve                  74178     8.3  1950    1950
Some Like It Hot              156432     8.3  1959    1950
The Killing                    56671     8.0  1956    1950
Roman Holiday                  87437     8.0  1953    1950
Touch of Evil                  65408     8.1  1958    1950
Rashômon                       90434     8.3  1950    1950
La strada                      42446     8.0  1954    1950
North by Northwest            198795     8.4  1959    1950
Sunset Blvd.                  123879     8.5  1950    1950
Shichinin no samurai          206216     8.7  1954    1950
12 Angry Men                  384187     8.9  1957    1950
Les diaboliques                36725     8.1  1955    1950
Kumonosu-jô                    26012     8.0  1957    1950
Ben-Hur                       141768     8.1  1959    1950
Witness for the Prosecution    53186     8.3  1957    1950
Dial M for Murder              92244     8.1  1954    1950
Ikiru                          36638     8.2  1952    1950
On the Waterfront              89233     8.2  1954    1950
Les quatre cents coups         61776     8.1  1959    1950
Paths of Glory                106038     8.4  1957    1950
High Noon                      72007     8.0  1952    1950
Vertigo                       218430     8.4  1958    1950
The Bridge on the River Kwai  132677     8.2  1957    1950
Rear Window                   280432     8.5  1954    1950
Det sjunde inseglet            98949     8.2  1957    1950
The Night of the Hunter        57974     8.0  1955    1950
Smultronstället                55861     8.2  1957    1950
Strangers on a Train           85012     8.1  1951    1950
Le salaire de la peur          31003     8.1  1953    1950

O que `imdb_por_nome[eh_de_1950s]` faz, precisamente, é percorrer `imdb_por_nome` linha por linha. Se a linha chamada *Singin' in the Rain* tiver o valor `True` em `eh_de_1950s`, essa linha será mantida. Se o valor for `False`, a linha será descartada. E assim por diante, para cada linha.

Observe que poderíamos ter conseguido isso sem nunca criar a variável `eh_de_1950s`, simplesmente colocando o código que usamos para criar a Series booleana diretamente dentro de `[...]`. Este é um padrão típico que você usará muito!

In [74]:
imdb_por_nome[imdb_por_nome.get('Decade') == 1950]

Votes  Rating  Year  Decade
Title                                                     
Singin' in the Rain           132823     8.3  1952    1950
All About Eve                  74178     8.3  1950    1950
Some Like It Hot              156432     8.3  1959    1950
The Killing                    56671     8.0  1956    1950
Roman Holiday                  87437     8.0  1953    1950
Touch of Evil                  65408     8.1  1958    1950
Rashômon                       90434     8.3  1950    1950
La strada                      42446     8.0  1954    1950
North by Northwest            198795     8.4  1959    1950
Sunset Blvd.                  123879     8.5  1950    1950
Shichinin no samurai          206216     8.7  1954    1950
12 Angry Men                  384187     8.9  1957    1950
Les diaboliques                36725     8.1  1955    1950
Kumonosu-jô                    26012     8.0  1957    1950
Ben-Hur                       141768     8.1  1959    1950
Witness for the Prosecution    53186     8.3  1957    1950
Dial M for Murder              92244     8.1  1954    1950
Ikiru                          36638     8.2  1952    1950
On the Waterfront              89233     8.2  1954    1950
Les quatre cents coups         61776     8.1  1959    1950
Paths of Glory                106038     8.4  1957    1950
High Noon                      72007     8.0  1952    1950
Vertigo                       218430     8.4  1958    1950
The Bridge on the River Kwai  132677     8.2  1957    1950
Rear Window                   280432     8.5  1954    1950
Det sjunde inseglet            98949     8.2  1957    1950
The Night of the Hunter        57974     8.0  1955    1950
Smultronstället                55861     8.2  1957    1950
Strangers on a Train           85012     8.1  1951    1950
Le salaire de la peur          31003     8.1  1953    1950

Ajuda ler os colchetes como "onde". Portanto, o comando na célula acima diz para manter todas as linhas de `imdb_por_nome` *onde* a década é a década de 1950.

Criar um novo DataFrame selecionando apenas certas linhas de um DataFrame existente que satisfaça alguma condição é chamado de *consulta*. A linha de código `imdb_por_nome[imdb_por_nome.get('Decade') == 1950]` é uma *consulta*.

**Questão 4.1.** Crie um DataFrame chamado `noventa_e_oito` contendo os filmes lançados em 1998.

In [75]:
noventa_e_oito = imdb_por_nome[imdb_por_nome.get('Year') == 1998]
noventa_e_oito

Votes  Rating  Year  Decade
Title                                                                   
Saving Private Ryan                         769893     8.5  1998    1990
American History X                          694602     8.5  1998    1990
Lock, Stock and Two Smoking Barrels (1998)  372863     8.2  1998    1990
The Big Lebowski                            473988     8.2  1998    1990
The Truman Show                             583004     8.0  1998    1990

Até agora só descobrimos onde uma coluna é *exatamente* igual a um determinado valor. No entanto, existem muitos outros operadores de comparação que poderíamos usar. Aqui estão alguns:

|Operador|Testes|
|-|-|
|`==`|a coisa da esquerda é igual à coisa da direita|
|`!=`|a coisa à esquerda *não* é igual à coisa à direita|
|`>`|a coisa à esquerda é maior que (e não igual) à coisa à direita|
|`>=`|a coisa à esquerda é maior ou igual à coisa à direita|
|`<`|a coisa à esquerda é menor que (e não igual) à coisa à direita|

As [notas de curso](https://notes.dsc10.com/02-data_sets/querying.html#examples) do DSC10 tem mais exemplos.

**Questão 4.2.** Utilizando os operadores da tabela acima, encontre todos os filmes com avaliação superior a 8,6. Coloque seus dados em um DataFrame chamado `realmente_bem_avaliados`.

In [76]:
realmente_bem_avaliados = imdb_por_nome[imdb_por_nome.get("Rating")>8.6]
realmente_bem_avaliados

Votes  Rating  Year  \
Title                                                                      
The Godfather                                      1027398     9.2  1972   
The Shawshank Redemption                           1498733     9.2  1994   
Il buono, il brutto, il cattivo (1966)              447875     8.9  1966   
The Lord of the Rings: The Two Towers               967389     8.7  2002   
The Dark Knight                                    1473049     8.9  2008   
Inception                                          1271949     8.7  2010   
Fight Club                                         1177098     8.8  1999   
The Godfather: Part II                              692753     9.0  1974   
Shichinin no samurai                                206216     8.7  1954   
Goodfellas                                          644556     8.7  1990   
12 Angry Men                                        384187     8.9  1957   
The Matrix                                         1073043     8.7  1999   
The Lord of the Rings: The Return of the King      1074146     8.9  2003   
Schindler's List                                    761224     8.9  1993   
Star Wars: Episode V - The Empire Strikes Back      700283     8.7  1980   
The Lord of the Rings: The Fellowship of the Ring  1099087     8.8  2001   
Star Wars                                           770011     8.7  1977   
One Flew Over the Cuckoo's Nest                     606395     8.7  1975   
Pulp Fiction                                       1166532     8.9  1994   
Forrest Gump                                       1078416     8.7  1994   

                                                   Decade  
Title                                                      
The Godfather                                        1970  
The Shawshank Redemption                             1990  
Il buono, il brutto, il cattivo (1966)               1960  
The Lord of the Rings: The Two Towers                2000  
The Dark Knight                                      2000  
Inception                                            2010  
Fight Club                                           1990  
The Godfather: Part II                               1970  
Shichinin no samurai                                 1950  
Goodfellas                                           1990  
12 Angry Men                                         1950  
The Matrix                                           1990  
The Lord of the Rings: The Return of the King        2000  
Schindler's List                                     1990  
Star Wars: Episode V - The Empire Strikes Back       1980  
The Lord of the Rings: The Fellowship of the Ring    2000  
Star Wars                                            1970  
One Flew Over the Cuckoo's Nest                      1970  
Pulp Fiction                                         1990  
Forrest Gump                                         1990

Qual é a maior avaliação de qualquer filme da década de 1990? Agora temos as ferramentas para responder a perguntas como essas. Dividindo em pedaços, encontramos primeiro todos os filmes da década de 1990:

In [77]:
eh_de_1990s = imdb_por_nome.get('Decade') == 1990
eh_de_1990s

Title
M                                        False
Singin' in the Rain                      False
All About Eve                            False
Léon                                      True
The Elephant Man                         False
                                         ...  
Forrest Gump                              True
Le salaire de la peur                    False
3 Idiots                                 False
Network                                  False
Eternal Sunshine of the Spotless Mind    False
Name: Decade, Length: 250, dtype: bool

Em seguida, selecionamos apenas estes filmes em nosso DataFrame:

In [78]:
de_1990s = imdb_por_nome[eh_de_1990s]
de_1990s

Votes  Rating  Year  Decade
Title                                                                    
Léon                                         635139     8.6  1994    1990
Mononoke-hime                                192165     8.4  1997    1990
Saving Private Ryan                          769893     8.5  1998    1990
In the Name of the Father                     95212     8.1  1993    1990
Before Sunrise                               158867     8.0  1995    1990
The Silence of the Lambs                     767224     8.6  1991    1990
The Shawshank Redemption                    1498733     9.2  1994    1990
Heat                                         388239     8.2  1995    1990
Toy Story                                    535249     8.3  1995    1990
Beauty and the Beast                         268480     8.0  1991    1990
Fight Club                                  1177098     8.8  1999    1990
L.A. Confidential                            376590     8.3  1997    1990
The Usual Suspects                           656756     8.6  1995    1990
Terminator 2: Judgment Day                   658564     8.5  1991    1990
Goodfellas                                   644556     8.7  1990    1990
American Beauty                              735056     8.4  1999    1990
The Sixth Sense                              630994     8.1  1999    1990
La haine                                      90937     8.0  1995    1990
Groundhog Day                                384272     8.0  1993    1990
Twelve Monkeys                               415809     8.0  1995    1990
The Matrix                                  1073043     8.7  1999    1990
Unforgiven                                   248514     8.3  1992    1990
American History X                           694602     8.5  1998    1990
Trois couleurs: Rouge                         57644     8.0  1994    1990
Reservoir Dogs                               578684     8.3  1992    1990
Casino                                       294394     8.2  1995    1990
Schindler's List                             761224     8.9  1993    1990
The Green Mile                               672878     8.5  1999    1990
Jurassic Park                                520391     8.0  1993    1990
Lock, Stock and Two Smoking Barrels (1998)   372863     8.2  1998    1990
Fargo                                        395997     8.1  1996    1990
Good Will Hunting                            529800     8.2  1997    1990
Trainspotting                                419372     8.1  1996    1990
Se7en                                        895411     8.6  1995    1990
Braveheart                                   653769     8.3  1995    1990
Underground                                   39447     8.0  1995    1990
The Big Lebowski                             473988     8.2  1998    1990
The Lion King                                548750     8.4  1994    1990
La vita è bella                              358305     8.6  1997    1990
The Truman Show                              583004     8.0  1998    1990
Pulp Fiction                                1166532     8.9  1994    1990
Forrest Gump                                1078416     8.7  1994    1990

Encontramos então a maior avaliação apenas destes filmes:

In [79]:
de_1990s.get('Rating').max()

9.2

Ou, se quiséssemos fazer tudo isso de forma mais concisa usando encadeamento:

In [80]:
imdb_por_nome[imdb_por_nome.get('Decade') == 1990].get('Rating').max()
#imdb_por_nome[imdb_por_nome.get('Decade')==1970].get('Rating').min()       -> Menor avaliacao da década de 1970

9.2

**Pergunta 4.3.** Encontre a avaliação média para filmes lançados no século 20 e a avaliação média para filmes lançados no século 21 para os filmes no `imdb`.

*Dica*: As séries possuem um método `.mean()`. Observe que o ano 2000 está no século 20 e que o filme mais antigo do conjunto de dados é de 1921!

In [81]:
avaliacao_media_do_seculo_20 = imdb_por_nome[(imdb_por_nome.get('Year') > 1920) & (imdb_por_nome.get('Year') <= 2000)].get('Rating').mean()
avaliacao_media_do_seculo_20

8.280113636363636

In [82]:
avaliacao_media_do_seculo_21 = imdb_por_nome[imdb_por_nome.get('Year')>2000].get('Rating').mean()
avaliacao_media_do_seculo_21

8.23108108108108

A propriedade `shape` informa quantas linhas e colunas existem em um DataFrame. (Uma "propriedade" é similar a um método que não precisa ser chamado adicionando parênteses.)

In [83]:
imdb_por_nome.shape

(250, 4)

Como um array, você pode obter o primeiro elemento do `shape` usando `[0]` e o segundo elemento usando `[1]`. Por exemplo, o número de linhas em `imdb_por_nome` é:

In [84]:
imdb_por_nome.shape[0]

250

In [85]:
imdb_por_nome.shape[1] #Colunas

4

Podemos usar isso para responder "Quantos filmes são do século 20?":

In [86]:
#imdb_por_nome[(imdb_por_nome.get('Year') > 1920) & (imdb_por_nome.get('Year') <= 2000)].shape[0]
imdb_por_nome[imdb_por_nome.get('Year') <= 2000].shape[0]

176

**Pergunta 4.4.** Use `shape` (e aritmética) para encontrar a *proporção* de filmes no conjunto de dados que foram lançados no século 20 e a proporção do século 21.

* **Dica:** A *proporção* de filmes lançados no século 20 é o *número* de filmes lançados no século 20, dividido pelo *número total* de filmes no conjunto de dados.

In [87]:
proporcao_do_seculo_20 = (imdb_por_nome[imdb_por_nome.get('Year') <= 2000].shape[0])/(imdb_por_nome.shape[0])
proporcao_do_seculo_20

0.704

In [88]:
proporcao_do_seculo_21 = (imdb_por_nome[imdb_por_nome.get('Year') > 2000].shape[0])/(imdb_por_nome.shape[0])
proporcao_do_seculo_21

0.296

**Pergunta 4.5.** Finalmente, vamos revisitar o DataFrame `populacao_por_ano` do início do laboratório. Calcule o ano em que a população mundial ultrapassou pela primeira vez os 7 mil milhões.

In [89]:
ano_que_a_populacao_ultrapassou_7_bilhoes = populacao_por_ano[populacao_por_ano.get('Populacao')>7000000000].index.min()
ano_que_a_populacao_ultrapassou_7_bilhoes

2011

# Linha de chegada

Parabéns! Você concluiu o Laboratório 1.

Para enviar sua tarefa:

1. Selecione `Kernel -> Reiniciar e executar tudo` para garantir que você executou todas as células.
2. Leia o caderno para ter certeza de que está tudo bem.
4. Baixe seu notebook usando `Arquivo -> Baixar como -> Notebook (.ipynb)` e, em seguida, carregue seu notebook para o Moodle.

In [90]:
#Thanks!